Tree icons & General plants fuckery & optimizations

comma opened the pull request:
Adds YUGE icons for exoplanet trees to use, with their support in icon generation.
Consolidates icon generation code for standalone plants and trays. Ugh.
Tree icons are ruined versions of https://opengameart.org/content/32x32-and-16x16-rpg-tiles-forest-and-some-interior-tiles by Stephen Challener and the Open Surge team ( http://opensnc.sourceforge.net), hosted by OpenGameArt.org

General plant cleanup.
Exoplanets now spawn fully grown plants directly instead of spawning vines - worked oddly for non-vine plants.
Exoplanets also have less different types of trees, for less silly looking forests.

comma committed:
Tree icons & General plants fuckery

Adds YUGE icons for exoplanet trees to use, with their support in icon generation.
Consolidates icon generation code for standalone plants and trays. Ugh.
Tree icons are ruined versions of https://opengameart.org/content/32x32-and-16x16-rpg-tiles-forest-and-some-interior-tiles by Stephen Challener and the Open Surge team ( http://opensnc.sourceforge.net), hosted by OpenGameArt.org

General plant cleanup.
Exoplanets now spawn fully grown plants directly instead of spawning vines - worked oddly for non-vine plants.
Exoplanets also have less different types of trees, for less silly looking forests.

PurpleMartinJCK commented:
Can we wait for #18914

comma commented:
never
There's barely any overlap

PurpleMartinJCK commented on code/modules/hydroponics/_hydro_setup.dm:
```diff
@@ -57,10 +57,17 @@
 #define TRAIT_BIOLUM_COLOUR        37
 #define TRAIT_IMMUTABLE            38
 #define TRAIT_FLESH_COLOUR         39
+#define TRAIT_LARGE            	   40
```
This doesn't look aligned 👀

comma replied:
![](https://vgy.me/rvYy6L.png)
I blame github. Open to suggestions on how to fix this dire issue to unblock the PR's merging.

PurpleMartinJCK replied:
👌 you're good dude.

thasc replied:
It's probably because there are a mixture of tabs and spaces in there. Replacing the tab with spaces would sort it and give everyone happy feelings inside.

PurpleMartinJCK commented on code/modules/hydroponics/_hydro_setup.dm:
```diff
@@ -57,10 +57,17 @@
+
+#define GROWTH_WORMS      		  "worms"
+#define GROWTH_VINES      		  "vines"
+#define GROWTH_BIOMASS         	  "mass"
```
This also doesn't look aligned (is it just github?)

comma replied:
![](https://vgy.me/i7lswA.png)

PurpleMartinJCK replied:
👌

PurpleMartinJCK commented on code/modules/overmap/exoplanets/exoplanet.dm:
```diff
@@ -319,21 +319,28 @@
+			S.chems["woodpulp"] = 1
+			big_flora_types += S
 
 /datum/random_map/noise/exoplanet/proc/spawn_flora(var/turf/T, var/big)
```
Was this that bit you were talking about before where you thought you used the plant effect wrong?

comma replied:
Yes, I wanted fully grown plants, and while spawning mature effect did the job, it was a very roundabout way of doing it.

PurpleMartinJCK commented:
Also can we get screenshots of how they look?

comma commented:
![](https://vgy.me/oE9gZf.png)

PurpleMartinJCK commented:
Cool, I like it.

comma committed:
Fixes rogue tree-vines

Jungle seed adaptation made trees spread like vines, which made them look really silly.

comma committed:
Shamanic optimizations

Axed sleep from vine processing.
Almost all of lag comes from update icon, so I concentrated on it.
Reduced calls to get_trait as those made a bulk of CPU time actually due to sheer number.
Reintroduced caching, actually useful at this scale.
Applied all overlays at once to avoid appearance churn.
Now trays check if they need to update icon isntead of doing it every tick (twice even)
Vines now can produce limited amount of children per plant.
Also made the child vines not spawn the proper plant, they stay vines forever.

comma commented:
Added some desperate monkey optimizations. Vines are more or less under control now, machinery is stll overloaded pretty heavily on lush worlds.

PurpleMartinJCK commented on code/modules/hydroponics/spreading/spreading_growth.dm:
```diff
@@ -78,7 +78,7 @@
 			health -= seed.handle_environment(T,T.return_air(),null,1)
 		if(health < max_health)
-			health += rand(3,5)
+			health += 1
```
Why make it grow more slowly?

comma replied:
Because even with controller slowdown they still grow almost every tick as you look at them.

PurpleMartinJCK commented on code/modules/hydroponics/spreading/spreading_growth.dm:
```diff
@@ -110,21 +110,15 @@
+			if(targets && targets.len && prob(round(seed.get_trait(TRAIT_POTENCY)/4)))
+				entangle(pick(targets))
+
+		if(parent.possible_children && neighbors.len && prob(spread_chance))
```
So only spreading to at most one neighbour now?

comma replied:
Per tick yes. It's called every tick, they can continue to spread to other neighbors after that.

comma commented:
Okay this definetly helps a bit.
Currently jungle planet with two vine seeds gets you 1000+ ms machinery ticks, (6000 ms vine ticks) 6 minutes into round, probably gets worse further on as vines settle into machinery trays.
This thing cuts that to 300 ms machinery / 2000 ms vine.
I am out of shitty hacks to cut it lower, it's problem of scale at this point, and would need more indepth solution later, but for now I believe this should help.

PurpleMartinJCK commented on code/modules/hydroponics/spreading/spreading_growth.dm:
```diff
@@ -78,11 +78,11 @@
 		if(health < max_health)
-			health += rand(3,5)
-			refresh_icon()
+			health += 1
```
How does it function stacked with #19032

comma replied:
p normal, I basically had those changes locally when I was testing.
It spreads fine

Techhead0 commented on code/modules/hydroponics/spreading/spreading.dm:
```diff
@@ -136,8 +130,38 @@
+		if(at_fringe >= spread_distance-3)
+			max_growth = max(2,max_growth-1)
+		if(at_fringe >= spread_distance-2)
+			max_growth = max(1,max_growth-1)
```
Low-priority, but this bit can probably just be done once in `Initialize()` instead of every `update_icon()`, can't it?

PurpleMartinJCK commented on code/modules/overmap/exoplanets/grass.dm:
```diff
@@ -29,11 +29,12 @@
+	if(!S.get_trait(TRAIT_LARGE))
+		var/vine_prob = rand(100)
+		if(vine_prob < 15)
+			S.set_trait(TRAIT_SPREAD,2)
```
If here you also made yield low, you'd limit the spread and speed of growth for exoplanet plants without impacting those on the torch.

comma replied:
Small stuff like that can be tweaked after we're out of lag abyss

PurpleMartinJCK replied:
Which is why I approved it anyway :P
